pilz_industrial_motion_testutils/src/pilz_industrial_motion_testutils/robot_motion_observer.py

```diff
@@ -0,0 +1,115 @@
+
+    The implementation is based on using the get_current_joint_values()) functionality of the RobotCommander().
+
+    :param group_name: Name of the planning group, default value is 'manipulator'
+    """
+
+    _DEFAULT_WAIT_TIME_FOR_MOTION_DETECTION = 3.0
+    _DEFAULT_TOLERANCE_FOR_MOTION_DETECTION = 0.01
+    _DEFAULT_SLEEP_INTERVAL = 0.01
+    _DEFAULT_GROUP_NAME = "manipulator"
+
+    def __init__(self, group_name=_DEFAULT_GROUP_NAME):
+        self._robot_commander = RobotCommander()
+        self._group_name = group_name
+
+    @staticmethod
+    def _detect_motion(joint_values_a, joint_values_b, tolerance=_DEFAULT_TOLERANCE_FOR_MOTION_DETECTION):
+        """TRUE if a significant motion was detected, otherwise FALSE."""
+        return not numpy.allclose(joint_values_a, joint_values_b,
+                                  atol=tolerance)
+
+    def _get_current_joint_states(self):
+        """Returns the current joint state values of the robot.
+        :raises RobotCurrentStateError if given planning group does not exist.
+        """
+        try:
+            return self._robot_commander.get_group(self._group_name).get_current_joint_values()
+        except MoveItCommanderException as e:
+            rospy.logerr(e.message)
+            raise RobotCurrentStateError(e.message)
+
+    def is_robot_moving(self,
+                        sleep_interval=_DEFAULT_SLEEP_INTERVAL,
+                        move_tolerance=_DEFAULT_TOLERANCE_FOR_MOTION_DETECTION):
+        """TRUE if the robot is currently moving, otherwise FLASE."""
+        start_position = self._get_current_joint_states()
+        rospy.sleep(sleep_interval)
+        return RobotMotionObserver._detect_motion(start_position, self._get_current_joint_states(), move_tolerance)
+
+    def wait_motion_start(self,
+                          wait_time_out=_DEFAULT_WAIT_TIME_FOR_MOTION_DETECTION,
+                          move_tolerance=_DEFAULT_TOLERANCE_FOR_MOTION_DETECTION,
+                          sleep_interval=_DEFAULT_SLEEP_INTERVAL):
+        """TRUE if the motion started in the given time interval, otherwise FALSE."""
+
+        old_joint_values = self._get_current_joint_states()
+        rospy.loginfo("Start joint values: " + str(old_joint_values))
+
+        motion_started = False
+        start_time = rospy.get_time()
+        rospy.loginfo("Wait until motion starts...")
+
+        while True:
+            rospy.sleep(sleep_interval)
+            curr_joint_values = self._get_current_joint_states()
+
+            if RobotMotionObserver._detect_motion(curr_joint_values, old_joint_values, move_tolerance):
+                motion_started = True
+                rospy.loginfo("Changed joint values detected: " + str(curr_joint_values))
+                rospy.loginfo("Motion started.")
+                break
+
+            if (rospy.get_time() - start_time > wait_time_out):
+                rospy.loginfo("Reached timeout when waiting for motion start.")
```

Comment:
To be really precise you should set `motion_started = False` to (better) cover edge cases.

Reply:
not necessary.